File: src/ai-analysis.ts
import { formatError } from './utils.js';

export interface AiConfig {
  provider: 'openai' | 'anthropic' | 'ollama';
  model: string;
  apiKey: string;
  prompt: string;
}

const DEFAULT_MODELS: Record<string, string> = {
  openai: 'gpt-4o-mini',
  anthropic: 'claude-haiku-4-5-20251001',
  ollama: 'llama3.2',
};

// B1 fix: Use promise chain for concurrency-safe rate limiting
const MIN_API_INTERVAL_MS = 1000;
let pending: Promise<void> = Promise.resolve();

/**
 * Run AI analysis on page content using the configured provider.
 * Requests are serialized via a promise chain to enforce rate limiting under concurrency.
 */
export function analyzeWithAi(
  config: AiConfig,
  pageContext: { url: string; title: string; description: string; h1: string; bodyText: string },
): Promise<string> {
  return new Promise<string>((resolve) => {
    pending = pending.then(async () => {
      await new Promise((r) => setTimeout(r, MIN_API_INTERVAL_MS));
      resolve(await doAnalyze(config, pageContext));
    });
  });
}

async function doAnalyze(
  config: AiConfig,
  pageContext: { url: string; title: string; description: string; h1: string; bodyText: string },
): Promise<string> {
  // Truncate body text to keep token usage reasonable
  const truncatedBody = pageContext.bodyText.substring(0, 2000);

  const userMessage = `${config.prompt}

Page URL: ${pageContext.url}
Title: ${pageContext.title}
Meta Description: ${pageContext.description}
H1: ${pageContext.h1}
Body Text (first 2000 chars): ${truncatedBody}`;

  const model = config.model || DEFAULT_MODELS[config.provider] || DEFAULT_MODELS.openai;

  try {
    switch (config.provider) {
      case 'openai':
        return await callOpenAi(config.apiKey, model, userMessage);
      case 'anthropic':
        return await callAnthropic(config.apiKey, model, userMessage);
      case 'ollama':
        return await callOllama(model, userMessage);
      default:
        return `Error: Unknown AI provider "${config.provider}"`;
    }
  } catch (err) {
    return `Error: ${formatError(err)}`;
  }
}

async function callOpenAi(apiKey: string, model: string, message: string): Promise<string> {
  const response = await fetch('https://api.openai.com/v1/chat/completions', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'Authorization': `Bearer ${apiKey}`,
    },
    body: JSON.stringify({
      model,
      messages: [
        { role: 'system', content: 'You are an SEO expert analyzing web pages. Be concise and actionable.' },
        { role: 'user', content: message },
      ],
      max_tokens: 500,
      temperature: 0.3,
    }),
    signal: AbortSignal.timeout(30000),
  });

  if (!response.ok) {
    const text = await response.text();
    throw new Error(`OpenAI API error: ${response.status} ${text.substring(0, 200)}`);
  }

  const data = await response.json();
  return data.choices?.[0]?.message?.content?.trim() || 'No response';
}

async function callAnthropic(apiKey: string, model: string, message: string): Promise<string> {
  const response = await fetch('https://api.anthropic.com/v1/messages', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'x-api-key': apiKey,
      'anthropic-version': '2023-06-01',
    },
    body: JSON.stringify({
      model,
      max_tokens: 500,
      messages: [
        { role: 'user', content: message },
      ],
      system: 'You are an SEO expert analyzing web pages. Be concise and actionable.',
    }),
    signal: AbortSignal.timeout(30000),
  });

  if (!response.ok) {
    const text = await response.text();
    throw new Error(`Anthropic API error: ${response.status} ${text.substring(0, 200)}`);
  }

  const data = await response.json();
  return data.content?.[0]?.text?.trim() || 'No response';
}

async function callOllama(model: string, message: string): Promise<string> {
  const response = await fetch('http://localhost:11434/api/generate', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      model,
      prompt: `You are an SEO expert analyzing web pages. Be concise and actionable.\n\n${message}`,
      stream: false,
    }),
    signal: AbortSignal.timeout(60000),
  });

  if (!response.ok) {
    const text = await response.text();
    throw new Error(`Ollama API error: ${response.status} ${text.substring(0, 200)}`);
  }

  const data = await response.json();
  return data.response?.trim() || 'No response';
}

/**
 * Resolve AI API key from explicit flag, env var, or fail.
 */
export function resolveAiKey(provider: string, explicitKey: string): string {
  if (explicitKey) return explicitKey;

  const envKeys: Record<string, string> = {
    openai: 'OPENAI_API_KEY',
    anthropic: 'ANTHROPIC_API_KEY',
  };

  const envName = envKeys[provider];
  if (envName) {
    const fromEnv = process.env[envName];
    if (fromEnv) return fromEnv;
    throw new Error(`No API key provided. Set ${envName} or use --ai-key`);
  }

  // Ollama doesn't need a key
  return '';
}

File: internal/analysis/aianalysis.go
package analysis

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"time"
)

const (
	aiMaxBodyLen    = 2000
	aiMaxTokens     = 500
	aiTemperature   = 0.3
	aiRateLimitMs   = 1000
	aiDefaultPrompt = "Analyze this page for SEO issues and opportunities. Provide 3-5 actionable recommendations."
	aiSystemPrompt  = "You are an SEO expert analyzing web pages. Be concise and actionable."
)

// AIProvider specifies the AI analysis provider.
type AIProvider string

const (
	AIOpenAI    AIProvider = "openai"
	AIAnthropic AIProvider = "anthropic"
	AIOllama    AIProvider = "ollama"
)

// AIConfig configures AI-powered page analysis.
type AIConfig struct {
	Provider AIProvider
	APIKey   string
	Model    string
	Prompt   string
}

// PageContext holds the relevant page data for AI analysis.
type PageContext struct {
	URL             string
	Title           string
	MetaDescription string
	H1              string
	BodyText        string
}

// ResolveAIKey resolves the API key from explicit value or environment.
func ResolveAIKey(provider AIProvider, explicitKey string) (string, error) {
	if explicitKey != "" {
		return explicitKey, nil
	}
	switch provider {
	case AIOpenAI:
		if key := os.Getenv("OPENAI_API_KEY"); key != "" {
			return key, nil
		}
		return "", fmt.Errorf("OPENAI_API_KEY environment variable not set")
	case AIAnthropic:
		if key := os.Getenv("ANTHROPIC_API_KEY"); key != "" {
			return key, nil
		}
		return "", fmt.Errorf("ANTHROPIC_API_KEY environment variable not set")
	case AIOllama:
		return "", nil // Ollama doesn't need a key
	}
	return "", fmt.Errorf("unknown AI provider: %s", provider)
}

// AnalyzeWithAI runs AI analysis on a page's content.
func AnalyzeWithAI(ctx context.Context, cfg AIConfig, page PageContext) (string, error) {
	if cfg.Model == "" {
		switch cfg.Provider {
		case AIOpenAI:
			cfg.Model = "gpt-4o-mini"
		case AIAnthropic:
			cfg.Model = "claude-haiku-4-5-20251001"
		case AIOllama:
			cfg.Model = "llama3.2"
		}
	}
	if cfg.Prompt == "" {
		cfg.Prompt = aiDefaultPrompt
	}

	bodyText := page.BodyText
	if len(bodyText) > aiMaxBodyLen {
		bodyText = bodyText[:aiMaxBodyLen]
	}

	message := fmt.Sprintf("%s\n\nPage URL: %s\nTitle: %s\nMeta Description: %s\nH1: %s\nBody Text (first %d chars): %s",
		cfg.Prompt, page.URL, page.Title, page.MetaDescription, page.H1, len(bodyText), bodyText)

	switch cfg.Provider {
	case AIOpenAI:
		return callOpenAIChat(ctx, cfg.APIKey, cfg.Model, message)
	case AIAnthropic:
		return callAnthropicChat(ctx, cfg.APIKey, cfg.Model, message)
	case AIOllama:
		return callOllamaGenerate(ctx, cfg.Model, message)
	}
	return "", fmt.Errorf("unknown AI provider: %s", cfg.Provider)
}

func callOpenAIChat(ctx context.Context, apiKey, model, message string) (string, error) {
	body := map[string]interface{}{
		"model": model,
		"messages": []map[string]string{
			{"role": "system", "content": aiSystemPrompt},
			{"role": "user", "content": message},
		},
		"max_tokens":  aiMaxTokens,
		"temperature": aiTemperature,
	}
	payload, err := json.Marshal(body)
	if err != nil {
		return "", fmt.Errorf("marshal OpenAI request: %w", err)
	}

	req, err := http.NewRequestWithContext(ctx, "POST", "https://api.openai.com/v1/chat/completions", bytes.NewReader(payload))
	if err != nil {
		return "", err
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", "Bearer "+apiKey)

	client := &http.Client{Timeout: 30 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	if resp.StatusCode != 200 {
		b, _ := io.ReadAll(io.LimitReader(resp.Body, 8192))
		return "", fmt.Errorf("OpenAI API error %d: %s", resp.StatusCode, string(b))
	}

	var result struct {
		Choices []struct {
			Message struct {
				Content string `json:"content"`
			} `json:"message"`
		} `json:"choices"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return "", err
	}
	if len(result.Choices) == 0 {
		return "", fmt.Errorf("no response from OpenAI")
	}
	return result.Choices[0].Message.Content, nil
}

func callAnthropicChat(ctx context.Context, apiKey, model, message string) (string, error) {
	body := map[string]interface{}{
		"model":      model,
		"max_tokens": aiMaxTokens,
		"system":     aiSystemPrompt,
		"messages": []map[string]string{
			{"role": "user", "content": message},
		},
	}
	payload, err := json.Marshal(body)
	if err != nil {
		return "", fmt.Errorf("marshal Anthropic request: %w", err)
	}

	req, err := http.NewRequestWithContext(ctx, "POST", "https://api.anthropic.com/v1/messages", bytes.NewReader(payload))
	if err != nil {
		return "", err
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("x-api-key", apiKey)
	req.Header.Set("anthropic-version", "2023-06-01")

	client := &http.Client{Timeout: 30 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	if resp.StatusCode != 200 {
		b, _ := io.ReadAll(io.LimitReader(resp.Body, 8192))
		return "", fmt.Errorf("Anthropic API error %d: %s", resp.StatusCode, string(b))
	}

	var result struct {
		Content []struct {
			Text string `json:"text"`
		} `json:"content"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return "", err
	}
	if len(result.Content) == 0 {
		return "", fmt.Errorf("no response from Anthropic")
	}
	return result.Content[0].Text, nil
}

func callOllamaGenerate(ctx context.Context, model, message string) (string, error) {
	prompt := aiSystemPrompt + "\n\n" + message
	body := map[string]interface{}{
		"model":  model,
		"prompt": prompt,
		"stream": false,
	}
	payload, err := json.Marshal(body)
	if err != nil {
		return "", fmt.Errorf("marshal Ollama request: %w", err)
	}

	req, err := http.NewRequestWithContext(ctx, "POST", "http://localhost:11434/api/generate", bytes.NewReader(payload))
	if err != nil {
		return "", err
	}
	req.Header.Set("Content-Type", "application/json")

	client := &http.Client{Timeout: 60 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	if resp.StatusCode != 200 {
		b, _ := io.ReadAll(io.LimitReader(resp.Body, 8192))
		return "", fmt.Errorf("Ollama API error %d: %s", resp.StatusCode, string(b))
	}

	var result struct {
		Response string `json:"response"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return "", err
	}
	return result.Response, nil
}
